bids/layout/writing.py

```diff
@@ -0,0 +1,180 @@
+    Replaces all entity names in a given pattern with the corresponding
+    values provided by entities.
+
+    Args:
+        entities (dict): A dictionary mapping entity names to entity values.
+        pattern (str): A path pattern that contains entity names denoted
+            by curly braces. Optional portions denoted by square braces.
+            For example: 'sub-{subject}/[var-{name}/]{id}.csv'
+            Accepted entity values, using regex matching, denoted within angle
+            brackets.
+            For example: 'sub-{subject<01|02>}/{task}.csv'
+
+    Returns:
+        A new string with the entity values inserted where entity names
+        were denoted in the provided pattern.
+    """
+    ents = re.findall(r'\{(.*?)\}', pattern)
+    new_path = pattern
+    for ent in ents:
+        match = re.search(r'([^|<]+)(<.*?>)?(\|.*)?', ent)
+        if match is None:
+            return None
+        name, valid, default = match.groups()
+        default = default[1:] if default is not None else default
+
+        if name in entities:
+            if valid is not None:
+                ent_val = str(entities[name])
+                if not re.match(valid[1:-1], ent_val):
+                    if default is None:
+                        return None
+                    entities[name] = default
+
+        ent_val = entities.get(name, default)
+        if ent_val is None:
+            return None
+        new_path = new_path.replace('{%s}' % ent, str(ent_val))
+
+    return new_path
+
+
+def build_path(entities, path_patterns, strict=False):
+    """
+    Constructs a path given a set of entities and a list of potential
+    filename patterns to use.
+
+    Args:
+        entities (dict): A dictionary mapping entity names to entity values.
+        path_patterns (str, list): One or more filename patterns to write
+            the file to. Entities should be represented by the name
+            surrounded by curly braces. Optional portions of the patterns
+            should be denoted by square brackets. Entities that require a
+            specific value for the pattern to match can pass them inside
+            carets. Default values can be assigned by specifying a string after
+            the pipe operator. E.g., (e.g., {type<image>|bold} would only match
+            the pattern if the entity 'type' was passed and its value is
+            "image", otherwise the default value "bold" will be used).
+                Example 1: 'sub-{subject}/[var-{name}/]{id}.csv'
+                Result 2: 'sub-01/var-SES/1045.csv'
+        strict (bool): If True, all passed entities must be matched inside a
+            pattern in order to be a valid match. If False, extra entities will
+            be ignored so long as all mandatory entities are found.
+
+    Returns:
+        A constructed path for this file based on the provided patterns.
+    """
+    if isinstance(path_patterns, string_types):
+        path_patterns = [path_patterns]
```

Comment:
I've seen this in a few places. Why not `listify`? Is there a distinction, or is it just that this existed in the thing that was being copied over?

Reply:
Good catch—it's vestigial from grabbit (I didn't write `build_path`). We should probably do a sweep and homogenize treatment.